File: leet_code/20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) <= 1 or len(s) % 2 != 0:
            return False

        tmp_s = ''
        for char in s:
            if char == '(' or char == '[' or char == '{':
                tmp_s += char
            elif len(tmp_s) == 0:
                return False
            elif char == ')' and tmp_s[-1] != '(' or char == ']' and tmp_s[-1] != '[' or char == '}' and tmp_s[-1] != '{':
                return False
            else:
                tmp_s = tmp_s[:-1]

        return len(tmp_s) == 0

File: leetcode/easy/20_valid_parentheses.cpp
// -----------------------------------------
// My Solution
//
// Time  Complexity: O(n)
// Space Complexity: O(n)
// -----------------------------------------
// n := s.size()
class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;
        for (int index = 0; index < s.size(); ++index) {
            if (s[index] == '(' || s[index] == '[' || s[index] == '{') {
                stack.push_back(s[index]);
            } else {
                if (stack.size() == 0)
                    return false;
                if (s[index] == ')' && stack.back() != '(' || s[index] == ']' && stack.back() != '[' || s[index] == '}' && stack.back() != '{')
                    return false;
                stack.pop_back();
            }
        }

        return stack.size() == 0;
    }
};
